src/gwt/src/org/rstudio/core/client/widget/events/ButtonClickManager.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * ButtonClickManager.java
+ *
+ * Copyright (C) 2009-19 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+package org.rstudio.core.client.widget.events;
+
+import com.google.gwt.dom.client.Document;
+import com.google.gwt.dom.client.NativeEvent;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.dom.client.DomEvent;
+import com.google.gwt.event.dom.client.KeyCodes;
+import com.google.gwt.event.dom.client.KeyDownEvent;
+import com.google.gwt.event.shared.HandlerManager;
+import com.google.gwt.event.shared.HasHandlers;
+import com.google.gwt.user.client.ui.Widget;
+
+/**
+ * Helper for extending UI elements that activate via click to also 
+ * activate via spacebar.
+ */
+public class ButtonClickManager extends HandlerManager implements HasHandlers
+{
+   public ButtonClickManager(Widget widget, ClickHandler clickHandler)
+   {
+      super(null);
+      widget.addDomHandler(clickHandler, ClickEvent.getType());
+      addHandler(ClickEvent.getType(), clickHandler);
+      widget.addDomHandler(event -> {
+            if (event.getNativeKeyCode() == KeyCodes.KEY_SPACE)
+            {
+               event.stopPropagation();
+               event.preventDefault();
+               click();
+            }
+         }, KeyDownEvent.getType()); 
+   }
```

Comment:
Since our app is single-paged this will likely never come up but I typically the standard in web pages is space and shift+space is page down/up and enter is for activating a selected item. How does that align with a11y standards?

Reply:
Good point, ARIA recommends both space and enter for clicking a button: https://www.w3.org/TR/wai-aria-practices/#button

Note that when focus is in a native HTML button, spacebar doesn't scroll, it only clicks the button.

I'm actually used to using spacebar to click buttons; enter, traditionally, activates the default button in a dialog (which isn't always the same button that has focus), but this model seems to be less common these days even in desktop apps.

Short answer, I'll look at adding support for both, but in a separate change. Thanks for the review!

Reply:
Looking elsewhere, we do have the "enter activates dialog" behavior in our code. For example, bring up the Global Options dialog, get focus on the "Browse" button (no visual indicator currently) and hitting Enter closes the dialog (equivalent to OK button), whereas spacebar activates the Browse button.